framework/src/main/java/org/checkerframework/framework/flow/CFAbstractValue.java

```diff
@@ -529,7 +574,23 @@ protected V upperBound(@Nullable V other, TypeMirror upperBoundTypeMirror, boole
             other.getUnderlyingType(),
             other.getAnnotations(),
             canBeMissingAnnotations(upperBoundTypeMirror));
-    return analysis.createAbstractValue(lub, upperBoundTypeMirror);
+    V upperBound = analysis.createAbstractValue(lub, upperBoundTypeMirror);
+    if (this.getThenStore() != null && other.getThenStore() != null) {
```

Comment:
What if one of `this` and `other` has a thenStore, but the other does not?  What should be done in that case?  The behavior is different than in `mostSpecific()` above, which always uses exactly one if either one is set.